Add `--normalize-case` option for snake_casing column names

acmiyaguchi opened the pull request:
This PR fixes #77 by adding a new option to `snake_case` all column names in a schema. This should be used by adding a `--normalize-case` flag to the command. By default, this option is turned off.

I've chosen [heck](https://crates.io/crates/heck) as the casing library, since it seems to have the largest number of active users. It uses the [`unicode_segmentation`](https://crates.io/crates/unicode-segmentation) crate to find word boundaries and performs `snake_casing` consistently across mixed casing. 

I've refactored the code to remove extra clones and to make the order of the functions flow better when reading top-down. I also added a few comments here and there.

acmiyaguchi commented:
I've updated the mozilla-pipeline-schema scripts to easily check the diff between different transpiler options.

I've created a diff of the `--normalize-case` option: https://gist.github.com/acmiyaguchi/3f526c440b67ebe469bcb6ab2da5123f

```bash
$ scripts/mps-generate-schemas.sh bq1 --type bigquery --resolve drop
...
80/132 succeded

$ scripts/mps-generate-schemas.sh bq2 --type bigquery --resolve drop --normalize-case
...
80/132 succeded

$ diff -q bq1/ bq2/
Files bq1/coverage.coverage.1.schema.json and bq2/coverage.coverage.1.schema.json differ
Files bq1/eng-workflow.hgpush.1.schema.json and bq2/eng-workflow.hgpush.1.schema.json differ
Files bq1/firefox-launcher-process.launcher-process-failure.1.schema.json and bq2/firefox-launcher-process.launcher-process-failure.1.schema.json differ
Files bq1/mozdata.event.1.schema.json and bq2/mozdata.event.1.schema.json differ
...

$ diff -q bq1/ bq2/ | wc -l
      45

$ diff bq1/ bq2/ > normalize_case.diff
```

acmiyaguchi commented on src/ast.rs:
```diff
@@ -1152,18 +1214,12 @@ mod tests {
+        assert_normalize(&tag, vec!["valid_name", "renamed_value_0", "_64bit"]);
+
+        tag.normalize_properties(true);
+        assert_normalize(&tag, vec!["valid_name", "renamed_value_0", "_64bit"]);
```
This case covers the behavior of the `heck` library, which will ignore punctuation outside of word boundaries e.g. `_64bit` is treated as `64bit`. This double checks that the casing library doesn't change this particular behavior.

badboy replied:
Maybe add that as a code comment? A PR comment and this knowledge will get lost

acmiyaguchi commented:
There are a couple of interesting cases from the diff that I want to highlight:

```
l2cacheKB -> l2cache_kb
speedMHz -> speed_m_hz
D2DEnabled -> d2d_enabled
DWriteEnabled -> d_write_enabled
activeGMPlugins -> active_gm_plugins
```

badboy commented on Cargo.toml:
```diff
@@ -13,6 +13,7 @@ name = "jst"
 [dependencies]
 clap = "~2.32"
 env_logger = "0.6.1"
+heck = "0.3.1"
```
👍

badboy commented on tests/normalize_case.rs:
```diff
@@ -0,0 +1,101 @@
+        r#"
+        [
+            {
+                "mode": "REQUIRED",
```
That the order in here differs from the definition above is just pure coincidence (because order in these definitions is irrelevant), right?

acmiyaguchi replied:
This is no coincidence, field names are sorted so schemas can be diff'ed in a deterministic way.

acmiyaguchi commented:
@badboy This PR has changed a bit from the last review, so I'm retagging you for review. We're looking to have a consistent implementation of snake-casing across the transpiler and ingestion, so I reimplemented the logic using regular expressions and string manipulation instead. This still maintains the same output as heck, but is portable to java and python3.

I've added 3 separate test cases to ensure that the behavior stays the same:
* alphanum_3 - all strings of length 3 drawn from the alphabet "aA7"
* word_4 - all strings of length 4 drawn from the alphabet "aA7_"
* mps-diff-integration - the strings that were generated by diffing mozilla-pipeline-schemas with this PR using heck

I also did the following:
* dropped the `regex` create in favor of `onig`, a wrapper around oniguruma for lookaround support
* made `to_snake_case` a function accessible via a public interface for testing.

jklukas commented on scripts/mps-generate-schemas.sh:
```diff
@@ -18,16 +18,18 @@ outdir=${1:-"avro"}
 for schema in $schemas; do
     namespace=$(basename $(dirname $(dirname $schema)))
-    schema_filename=$(basename $schema | sed 's/schema.json/avro.json/g')
+    schema_filename=$(basename $schema)
```
I think this is the point where we need to normalize casing of namespace and doctype names. Or perhaps we just have a special case to modify `untrustedModules` to `untrusted_modules`, and we create an issue in mps to fail a test when adding a new schema with capital letters in the name.

whd replied:
I think the renaming needs to happen [here](https://github.com/mozilla/mozilla-schema-generator/blob/master/bin/schema_generator.sh#L148) as well since these scripts are AFAIK unused by MSG.

Having MPS fail a test when adding other pings with this property sounds swell.

jklukas replied:
If this script is just used for testing within this repo, then we don't need to worry about snake-casing here. I'd agree that it's the script in MSG where the change needs to happen.

jklukas replied:
Filed https://github.com/mozilla-services/mozilla-pipeline-schemas/issues/355

acmiyaguchi replied:
Yep, this script is the prototype of the script in [`gcp-ingestion/ingestion-beam`](https://github.com/mozilla/gcp-ingestion/blob/master/ingestion-beam/bin/generate-avro-schemas) and is mostly for end-to-end testing of encoding json into avro, then importing into bigquery and reading the schema back. The transpiler reads from a file (or stdin), so it doesn't have a bias on how the schema repo is organized.

I'm going to leave this as is, for conciseness.

jklukas replied:
I am going to prep a PR for MGS

jklukas replied:
See https://github.com/mozilla/mozilla-schema-generator/pull/46

badboy commented on src/casing.rs:
```diff
@@ -0,0 +1,61 @@
+/// detected by a lowercase followed by an uppercase. Numbers can take on either
+/// case depending on the preceeding symbol.
+///
+/// See: https://github.com/withoutboats/heck/blob/master/src/lib.rs#L7-L17
```
Maybe better to use a fixed-point-in-time link: https://github.com/withoutboats/heck/blob/093d56fbf001e1506e56dbfa38631d99b1066df1/src/lib.rs#L7-L17

badboy commented on src/casing.rs:
```diff
@@ -0,0 +1,61 @@
+            \b                              # standard word boundary
+            |(?<=[a-z][A-Z])(?=\d*[A-Z])    # break on runs of uppercase e.g. A7Aa -> A7|Aa
+            |(?<=[a-z][A-Z])(?=\d*[a-z])    # break in runs of lowercase e.g a7Aa -> a7|Aa
+            |(?<=[A-Z])(?=\d*[a-z])         # ends with an uppercase e.g. a7A -> a7|A
```
I'm a regex fan, but ... oh boy :D 
Did you come up with the regex here or is that copied from somewhere?

acmiyaguchi replied:
I [came up with the regex](https://github.com/acmiyaguchi/test-casing/blob/8ca3d68db512fd3a17868c0b08cc84909ebebbc7/src/main.py#L1-L34) in a separate project for testing all of the different casing functions/libraries that we've been using. Setting up the tests was the key result, I have an [auto-generated reporting](https://github.com/acmiyaguchi/test-casing/blob/master/RESULTS.md) for different implementations of snake_casing, including this regex algorithm that's implemented in python.

Here are some other resources that I ended up using:

* [[RexEgg] - Regex Boundaries and Delimiters—Standard and Advanced](https://www.rexegg.com/regex-boundaries.html) - This introduced the concept of word boundaries like `\b` and how to implement it from scratch using lookaheads and lookbehinds.
* [[StackOverflow] - RegEx to split camelCase or TitleCase (advanced)](https://stackoverflow.com/a/7599674) - This validated my approach as I was developing the regex in an online REPL. It does not include digits, which can be either uppercase or lowercase depending on the previous letter.
* [[StackOverflow] - What's the technical reason for “lookbehind assertion MUST be fixed length” in regex?](https://stackoverflow.com/a/40078049) - So how can you tell whether a digit is lowercase or uppercase (e.g. `f("a7aA") == f("A7AA")`)? My first attempt was something similar to `(?<=[a-z]\d*)(?=[A-Z][a-z])` for a "lowercase" digit, but this is not supported by most engines because dynamic look-behinds can perform very badly. I found a neat idea for "inverted" positive lookaheads, which can be used to figure out if a digit is lowercase or uppercase in the 2nd and 3rd lines of the regex.

These three resources, along with the definition of a word boundary and the reference test cases in heck, were useful with developing the regex.

badboy commented on src/ast.rs:
```diff
@@ -332,60 +322,176 @@ impl Tag {
 
-    fn fill_names(&mut self, name: String, namespace: String) {
-        self.name = Some(name);
+    /// Get the path the the current tag in the context of the larger schema.
```
```suggestion
    /// Get the path to the current tag in the context of the larger schema.
```

badboy commented on src/ast.rs:
```diff
@@ -332,60 +322,176 @@ impl Tag {
+    ///
+    /// The fully qualified names are as follows:
+    ///
+    /// * `root.attributes.timestamp`
```
```suggestion
    /// * `root.environment.timestamp`
```

badboy commented on src/ast.rs:
```diff
@@ -332,60 +322,176 @@ impl Tag {
+                // Replace property names with the normalized property name
+                if let Some(mut renamed) = Tag::normalize_name_bigquery(&key) {
+                    renamed = if normalize_case {
+                        // heck::SnakeCase will strip all punctuation outside of word boundaries.
```
Left-over comment?

acmiyaguchi replied:
Yep, definitely a stale comment.

whd commented:
We're going to need a version bump for this new option so that we can reference it in [MSG](https://github.com/mozilla/mozilla-schema-generator/blob/master/Dockerfile#L27).
